blockchain/blockchain.go

```diff
@@ -400,6 +400,10 @@ func blockByNumber(txn db.Transaction, number uint64) (*core.Block, error) {
 		block.Receipts = append(block.Receipts, receipt)
 	}
 
+	if err := iterator.Error(); err != nil {
```

Comment:
I agree this doesn't match the previous code. As far as I can tell, achieving equivalence would require 

1. changing `return nil, err` statements to something like `return nil, db.CloseAndWrapOnErr(closeFn, err)`
2. "manually" closing the resource and checking for error before returning from the happy path.

To avoid these code smells, we must always release the resource without error and catch any "internal database errors" before returning from the happy path. `defer resource.Discard()` and `resource.Error()` accomplish these respective goals.

I agree it's not a great solution, but I'm not sure how else to separate releasing resources from error handling. I'm definitely open to alternatives.